File: №23 Visitor Pattern/Python/main.py
from abc import ABC, abstractmethod

class Visitor(ABC):
    @abstractmethod
    def visit(self, movie_booking):
        pass

class BookingAnalyzer(Visitor):
    def visit(self, movie_booking):
        print(f"Analyzing movie booking: {movie_booking}")

class Element(ABC):
    @abstractmethod
    def accept(self, visitor):
        pass

class MovieBooking(Element):
    def __init__(self, movie_name, number_of_tickets):
        self.movie_name = movie_name
        self.number_of_tickets = number_of_tickets

    def accept(self, visitor):
        visitor.visit(self)

    def get_movie_name(self):
        return self.movie_name

    def get_number_of_tickets(self):
        return self.number_of_tickets

    def __str__(self):
        return f"Movie: {self.movie_name}, Tickets: {self.number_of_tickets}"

if __name__ == "__main__":
    movie_booking1 = MovieBooking("Avengers: Endgame", 3)
    movie_booking2 = MovieBooking("Spider-Man: No Way Home", 5)

    analyzer = BookingAnalyzer()
    movie_booking1.accept(analyzer)
    movie_booking2.accept(analyzer)

File: №23 Visitor Pattern/Java/src/Main.java

interface Visitor {
    void visit(MovieBooking movieBooking);
}


class BookingAnalyzer implements Visitor {
    @Override
    public void visit(MovieBooking movieBooking) {
        System.out.println("Analyzing movie booking: " + movieBooking);
    }
}

interface Element {
    void accept(Visitor visitor);
}

class MovieBooking implements Element {
    private final String movieName;
    private final int numberOfTickets;

    public MovieBooking(String movieName, int numberOfTickets) {
        this.movieName = movieName;
        this.numberOfTickets = numberOfTickets;
    }

    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    public String getMovieName() {
        return movieName;
    }

    public int getNumberOfTickets() {
        return numberOfTickets;
    }

    @Override
    public String toString() {
        return "Movie: " + movieName + ", Tickets: " + numberOfTickets;
    }
}

public class Main {
    public static void main(String[] args) {
        MovieBooking movieBooking1 = new MovieBooking("Avengers: Endgame", 3);
        MovieBooking movieBooking2 = new MovieBooking("Spider-Man: No Way Home", 5);

        Visitor analyzer = new BookingAnalyzer();
        movieBooking1.accept(analyzer);
        movieBooking2.accept(analyzer);
    }
}
